### ISSUE TITLE
300+ videos in a course, last 70+ videos problem

### ISSUE THREAD

**Issue by Galibri** [FaisalUmair/udemy-downloader-gui#183]
I have a course of over 330+ vidoes and files. It's downloading almost all the videos. But for the last 70+ videos, The files are not encoded in right way to play and they have only few bytes after downloading.

Other courses are fine and working well.

**Comment by intermediatech** [comment]
Having same issue

**Comment by andiwinter** [comment]
The content of the files not correctly downloaded is always the same: an authorization error, instead of file data. Maybe you can detect this error and do a retry.

<HTML><HEAD>
<TITLE>401 Authorization Required</TITLE>
</HEAD>
<BODY>
<H1>Authorization Required</H1>
This server could not verify that you
are authorized to access the document you
requested.  Either you supplied the wrong
credentials (e.g., bad password), or your
browser doesn't understand how to supply
the credentials required.<P>
</BODY></HTML>

**Comment by andiwinter** [comment]
Sorry, HTML content is automatically formatted, therefore see attached sample
[sample.zip](https://github.com/FaisalUmair/udemy-downloader-gui/files/2563125/sample.zip)

**Comment by FaisalUmair** [comment]
@Galibri @intermediatech I believe the problem is due to the download links getting expired because of time taken to download the initial lectures. 

@andiwinter You are right. As I mentioned the download links get expired, we can detect the error and retry download with fresh links.

**Comment by Galibri** [comment]
@FaisalUmair I have taken the number of last downloaded video and started with the next one to the last one in a second phase. That works for me.

**Comment by FaisalUmair** [comment]
@Galibri Yes, making use of Download Start/ Download End seems to be a workaround for now.

@andiwinter The fact that your internet connection is slow makes my point even stronger. If your connection is slow then obviously it will take time for the downloader to reach the ending lectures of the course, there is a chance that when the downloader starts downloading the last few lectures of the course, their download links are expired.

And yes, I do agree with you that there should be a download queue feature where a user can select how many courses to download at once. I will consider adding this feature in upcoming releases of the program.
